gateway/gateway.go

```diff
@@ -56,32 +73,60 @@ func newDTLSListener(ctx context.Context, cfg *GatewayConfig, address *net.UDPAd
 	var certificate *tls.Certificate
 	var err error
 
-	if cfg.SelfSigned {
-		var cert tls.Certificate
-		cert, err = selfsign.GenerateSelfSigned()
-		certificate = &cert
-	} else {
-		privateKey := cfg.PrivateKey
-		if privateKey == nil {
-			err = errors.New("private key is missing")
-		}
-		if certificate = cfg.Certificate; certificate != nil {
-			certificate.PrivateKey = privateKey
+	logger := util.NewProductionLogger("gateway")
+
+	if !cfg.UsePSK && cfg.UseDTLS {
+		if cfg.SelfSigned {
+			var cert tls.Certificate
+			cert, err = selfsign.GenerateSelfSigned()
+			certificate = &cert
 		} else {
-			err = errors.New("TLS certificate is missing")
+			privateKey := cfg.PrivateKey
+			if privateKey == nil {
+				err = errors.New("private key is missing")
+			}
+			if certificate = cfg.Certificate; certificate != nil {
+				certificate.PrivateKey = privateKey
+			} else {
+				err = errors.New("TLS certificate is missing")
+			}
 		}
 	}
+
 	if err != nil {
 		return nil, err
 	}
 
 	dtlsConfig := &dtls.Config{
-		Certificates:         []tls.Certificate{*certificate},
 		ExtendedMasterSecret: dtls.RequireExtendedMasterSecret,
 		ConnectContextMaker: func() (context.Context, func()) {
 			return context.WithTimeout(ctx, dtlsConnectTimeout)
 		},
 	}
+
+	if !cfg.UsePSK && cfg.UseDTLS && certificate != nil {
+		dtlsConfig.Certificates = []tls.Certificate{*certificate}
+	}
+
+	if cfg.UsePSK && cfg.UseDTLS {
+		dtlsConfig.CipherSuites = []dtls.CipherSuiteID{dtls.TLS_PSK_WITH_AES_128_GCM_SHA256}
+		dtlsConfig.PSK = func(hint []byte) ([]byte, error) {
+			psk, ok := cfg.PSK.Get(string(hint))
+			if ok {
+				return psk.([]byte), nil
+			}
+
+			psk, ok = getPSK(string(hint), cfg, logger)
+			if ok {
+				cfg.PSK.Set(string(hint), psk, cfg.PSKCacheExpiration)
+				return psk.([]byte), nil
+			}
+
+			return nil, errors.New("PSK not found")
+		}
+		dtlsConfig.PSKIdentityHint = []byte(cfg.PSKIdentity)
+	}
+
```

Comment:
This code duplicates the `client.go` you added. Can't we share it? PSK "client"?